ubi_manifest/app/api.py

```diff
@@ -1,8 +1,150 @@
 
 router = APIRouter(prefix="/api/v1")
 
 
 @router.get("/status")
 async def status():
     return {"status": "OK"}
+
+
+@router.post(
+    "/manifest",
+    response_model=List[TaskState],
+    status_code=201,
+    responses={
+        201: {
+            "description": "Depsolve tasks created",
+            "content": {
+                "application/json": {
+                    "example": [
+                        {
+                            "task_id": "some_task_id",
+                            "state": "PENDING",
+                        }
+                    ]
+                }
+            },
+        },
+        404: {
+            "description": "None of repositories requested are not allowed for depsolving.",
+            "content": {
+                "application/json": {
+                    "example": {
+                        "details": "None of [repo_id] are allowed for depsolving."
+                    }
+                }
+            },
+        },
+    },
+)
+async def manifest_post(depsolve_item: DepsolveItem) -> List[TaskState]:
+    repo_groups = {}
+    # compare provided repo_ids with the config and pick allowed repo groups
+    for repo_id in depsolve_item.repo_ids:
+        for key, group in app.conf.allowed_ubi_repo_groups.items():
+            if repo_id in group:
+                repo_groups.setdefault(key, group)
+
+    if not repo_groups:
+        raise HTTPException(
+            status_code=404,
+            detail=f"None of {depsolve_item.repo_ids} are allowed for depsolving.",
+        )
+
+    tasks_states = []
+    for repo_group in repo_groups.values():
+        task = depsolve_task.apply_async(args=[repo_group])
+        tasks_states.append(TaskState(task_id=task.task_id, state=task.state))
+
+    return tasks_states
+
+
+@router.get(
+    "/manifest/{repo_id}",
+    response_model=DepsolverResult,
+    status_code=200,
+    responses={
+        200: {
+            "description": "Depsolved content for repo_id found",
+            "content": {
+                "application/json": {
+                    "example": {
+                        "repo_id": "foo-bar-repo",
+                        "content": [
+                            {
+                                "src_repo_id": "source-foo-bar-repo",
+                                "unit_type": "RpmUnit",
+                                "unit_attr": "filename",
+                                "value": "some-filename.rpm",
+                            }
+                        ],
+                    }
+                }
+            },
+        },
+        404: {
+            "description": "Content for request repository is not available.",
+            "content": {
+                "application/json": {
+                    "example": {"detail": "Content for foo-repo not found"}
+                }
+            },
+        },
+    },
+)
+async def manifest_get(repo_id: str) -> DepsolverResult:
```

Comment:
Am I understanding right the general API usage:

- POST to /manifest with the repo_ids you care about.  You'll get some task IDs back.
- poll /task/{task_id} until your task(s) are done.
- Then you can GET /manifest/{repo_id} for each repo, and you know that the data is up-to-date.

And also some subtle things:

- If you don't care whether you get a 100% up-to-date manifest, you could maybe query /manifest/{repo_id} without bothering to POST to /manifest first.  (Not clear if anyone would want to do that)
- The state of the manifest for each repo is shared across all clients.  So for example if client A and B both POST to /manifest with the same repo_ids around the same time, and client A's tasks complete first, A can be fetching manifests for repos; and then the manifests might change again as B's tasks complete.  So that's something the client has to be aware of: the manifest can change at any time without warning.

Reply:
Yeah, the basic usage would be:
* POST to /manifest, poll /task/{task_id}, GET to /manifest/{repo_id}
But I was thinking also about scenario "not 100% up-to-date manifest":
* GET to /manifest/{repo_id}
** if found - take the data
** else: do POST /manifest, poll /task and GET /manifest/{repo_id}
- manifest should be always fresh enough - by default settings any manifest stored in redis will live only 4 hours.

About the:
> The state of the manifest for each repo is shared across all clients.

It doesn't make sense to run depsolving for one repo multiple times at one moment. I was thinking about some locking mechanism.
- if there is a lock for some repo_id, do not enqueue new depsolving task.
- I think this is something we may consider in future.

Also we can consider to create some versioning mechanism for manifests - e.g. store N latest manifests.
It may be useful for auditing.